program/rust/src/time_machine_types.rs

```diff
@@ -1,45 +1,192 @@
+    pub slot_gap:            u64,
+    pub price:               i64,
 }
+
 impl PriceAccountWrapper {
     pub fn initialize_time_machine(&mut self) -> Result<(), OracleError> {
-        // TO DO
+        #[cfg(test)]
+        self.time_machine
+            .initialize(THIRTY_MINUTES, PC_MAX_SEND_LATENCY.into());
         Ok(())
     }
 
     pub fn add_price_to_time_machine(&mut self) -> Result<(), OracleError> {
-        // TO DO
+        #[cfg(test)]
+        self.time_machine.add_datapoint( &DataPoint{
+            last_two_timestamps : (self.price_data.prev_timestamp_, self.price_data.timestamp_),
+            slot_gap : (self.price_data.last_slot_ - self.price_data.prev_slot_),
+            price: self.price_data.prev_price_ /2  + self.price_data.agg_.price_ / 2 + (self.price_data.prev_price_ % 2) * (self.price_data.agg_.price_ % 2), // Hack to avoid overflow?
+        }
+        )?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+pub struct SmaTracker<const NUM_BUCKETS: usize> {
+    /// The maximum gap in slots for an epoch's sma to be valid, we set it to PC_MAX_SEND_LATENCY
+    pub threshold:                     u64,
+    /// The legth of one sma epoch.
+    pub granularity:                   i64,
+    /// Numerator for the current epoch (ex : slot_gap1 * price_1 + slot_gap2 * price_2 + slot_gap3
+    /// * price_3)
+    /// It can never overflow because :
+    /// slot_gap1 * price_1 + slot_gap2 * price_2 + slot_gap3 * price_3 <= (slot_gap1 + slot_gap2 +
+    /// slot_gap3) * i64::MAX <= clock_slot * i128::MAX
+    pub current_epoch_numerator:       i128,
+    /// Denominator for the current epoch (ex : slot_gap1 + slot_gap2 + slot_gap3)
+    /// It can never overflow because : (slot_gap1 + slot_gap2 + slot_gap3) <= clock_slot
+    pub current_epoch_denominator:     u64,
+    /// Whether the current epoch is valid (i.e. max slot_gap <= threshold) for all the epochs
+    /// updates
+    pub current_epoch_is_valid:        bool,
+    /// Stores the running sum of individual epoch averages, we can support u64::MAX epochs. And
+    /// clock.unix_timestamp <= u64::MAX so the clock will break first.
+    pub running_sum_of_price_averages: [i128; NUM_BUCKETS],
+    /// Stores the number of valid epochs since inception, we can support u64::MAX epochs. And
+    /// clock.unix_timestamp <= u64::MAX so the clock will break first.
+    pub running_valid_epoch_counter:   [u64; NUM_BUCKETS],
+}
+
+impl<const NUM_BUCKETS: usize> SmaTracker<NUM_BUCKETS> {
+    pub fn time_to_epoch(&self, time: i64) -> Result<usize, OracleError> {
+        try_convert::<i64, usize>(time / self.granularity) // Can never fail because usize is u64
+                                                           // and time is positive
+    }
+
+    pub fn initialize(&mut self, granularity: i64, threshold: u64) {
+        self.threshold = threshold;
+        self.granularity = granularity;
+        self.current_epoch_is_valid = true;
```

Comment:
shouldn't this start as `false` because we're initializing in the middle of an epoch?

Reply:
I think you're right